src/main/lighting/SpotLight.java

```diff
@@ -0,0 +1,91 @@
+package lighting;
+
+import math.compare.DoubleCompare;
+import primitives.Colour;
+import primitives.NormalizedVector;
+import primitives.Point;
+
+/**
+ * This class is represents an directional light source that attenuates with
+ * respect to distance.
+ *
+ *
+ * @author Abraham Murciano
+ * @author Eli Levin
+ */
+public class SpotLight extends PointLight {
+
+	private NormalizedVector direction;
+	private double divergence;
+	private double sharpness;
+
+	/**
+	 * Construct a {@link SpotLight} from a {@link Colour}, a direction, and 3
+	 * doubles that represent the attenuation constants.
+	 * 
+	 * @param colour    The colour of the light
+	 * @param position  The position in space of the light source
+	 * @param direction The direction of the spot light
+	 * @param q         The quadratic factor
+	 * @param l         The linear factor
+	 * @param c         The constant factor
+	 */
+	public SpotLight(Colour colour, Point position, double q, double l, double c, NormalizedVector direction) {
+		super(colour, position, q, l, c);
+		this.direction = direction;
+		this.divergence = this.sharpness = 1;
+	}
+
+	/**
+	 * Construct a {@link SpotLight} from a {@link Colour}, a direction, and 3
+	 * doubles that represent the attenuation constants.
+	 * 
+	 * @param colour     The colour of the light
+	 * @param position   The position in space of the light source
+	 * @param q          The quadratic factor
+	 * @param l          The linear factor
+	 * @param c          The constant factor
+	 * @param direction  The direction of the spot light
+	 * @param divergence The divergence of the spot light
+	 * @param sharpness The sharpness of the spot light
+	 */
+	public SpotLight(Colour colour, Point position, double q, double l, double c, NormalizedVector direction,
+			double divergence, double sharpness) {
+		this(colour, position, q, l, c, direction);
+		this.direction = direction;
+		if (divergence <= 1) {
+			throw new IllegalArgumentException("Divergence must be greater than 1.");
+		}
+		this.divergence = divergence;
+		if (sharpness <= 0) {
+			throw new IllegalArgumentException("Sharpness must be greater than 0.");
+		}
+		this.sharpness = sharpness;
+
+	}
+
+	/**
+	 * Calculate the colour of the light that this light source is projecting onto a
+	 * given point.
+	 *
+	 * @param point The point at which to calculate the colour of the light.
+	 * @return The colour of the light at the given point.
+	 */
+	@Override
+	public Colour colourAt(Point point) {
+		double cosA = direction.dot(vectorTo(point));
+		double scalar = ((DoubleCompare.neq(divergence, 1) && DoubleCompare.neq(sharpness, 1)) ? directionalFactor(cosA)
```

Comment:
Is it really worth the check? Perhaps this indicates that it may be suitable to have two kinds of spotlight. The basic one and a `NarrowSpotlight extends Spotlight` which will override the colour at function.

Or better yet, you can instead of oversizing colourAt, have a protected function
```java
double scalar(Point point) {
    return direction.dot(vectotTo(point));
}
```

And in NarrowSpotlight,
```java
@Override
double scalar(Point point) {
    return directionalFactor(super.scalar(point));
}
```

Then you can simply call this function in colourAt instead of calculating the scalar there